app/contracts/work_packages/base_contract.rb

```diff
@@ -335,28 +347,111 @@ def validate_remaining_work_is_lower_than_work
       end
     end
 
+    # to be removed in 15.0 with :percent_complete_edition feature flag removal
     def validate_remaining_work_is_set_when_work_is_set
       if work_set? && !remaining_work_set?
         errors.add(:remaining_hours, :must_be_set_when_work_is_set)
       end
     end
 
+    # to be removed in 15.0 with :percent_complete_edition feature flag removal
     def validate_work_is_set_when_remaining_work_is_set
       if remaining_work_set? && !work_set?
         errors.add(:estimated_hours, :must_be_set_when_remaining_work_is_set)
       end
     end
 
+    def validate_work_is_set_when_remaining_work_and_percent_complete_are_set
+      if remaining_work_set_and_valid? && percent_complete_set_and_valid? && work_unset?
+        errors.add(:estimated_hours, :must_be_set_when_remaining_work_and_percent_complete_are_set)
+      end
+    end
+
+    def validate_remaining_work_is_set_when_work_and_percent_complete_are_set
+      if work_set_and_valid? && percent_complete_set_and_valid? && remaining_work_unset?
+        errors.add(:remaining_hours, :must_be_set_when_work_and_percent_complete_are_set)
+      end
+    end
+
+    def validate_percent_complete_is_set_when_work_and_remaining_work_are_not_set
+      if work_set_and_valid? && remaining_work_set_and_valid? && work != 0 && percent_complete_unset?
+        errors.add(:done_ratio, :must_be_set_when_work_and_remaining_work_are_set)
+      end
+    end
+
+    # rubocop:disable Metrics/AbcSize
+    def validate_percent_complete_matches_work_and_remaining_work
+      return if WorkPackage.status_based_mode? || percent_complete_unset? || work == 0
+      return if remaining_work_exceeds_work? # avoid too many error messages at the same time
+      return unless work_set? && remaining_work_set?
+
+      work_done = work - remaining_work
+      expected_percent_complete = (100 * work_done.to_f / work).round
+
+      if percent_complete != expected_percent_complete
+        errors.add(:done_ratio, :does_not_match_work_and_remaining_work)
+      end
+    end
+    # rubocop:enable Metrics/AbcSize
+
+    def validate_percent_complete_is_unset_when_work_is_zero
+      return if WorkPackage.status_based_mode?
+
+      if work == 0 && percent_complete_set?
+        errors.add(:done_ratio, :cannot_be_set_when_work_is_zero)
+      end
+    end
+
+    def work
+      model.estimated_hours
+    end
+
     def work_set?
       model.estimated_hours.present?
     end
 
+    def work_set_and_valid?
+      work_set? && work >= 0
+    end
+
+    def work_unset?
+      model.estimated_hours.nil?
+    end
```

Comment:
I like this "stacking" style on these methods as each one builds from the previous. Do you think as

```ruby
def work
  model.estimated_hours
end
```

is defined, the other 3 methods in the set should reference `#work` and not `model.estimated_hours` again?

Reply:
You're right, I'll fix that.